### PR TITLE
MCS-3968 Hashkey FTL for log creation

### PR DESCRIPTION
**Pull request by javierscope**

### CODE REVIEW

File: src/main/java/com/scoperetail/fusion/core/application/service/command/PosterService.java [hunk 1]
```diff
@@ -83,6 +85,35 @@ public void post(final String event, final Object domainEntity, final boolean is
...
+      if (optConfig.isPresent()) {
+        final Config config = optConfig.get();
+        final Optional<Adapter> optAdapter = config.getAdapters().stream()
+            .filter(a -> !a.transformationType.equals(NONE)).findFirst();
+        if (optAdapter.isPresent()) {
+          final Adapter adapter = optAdapter.get();
+          Map<String, Object> paramsMap = new HashMap<>();
+          paramsMap.put(Transformer.DOMAIN_ENTITY, domainEntity);
+          if(adapter.getTransformationType() == TransformationType.DOMAIN_EVENT_FTL_TRANSFORMER)
+            return domainToDomainEventJsonFtlTransformer.getHashKey(event, paramsMap, adapter.getTemplate());
```
Comment (amit-petkar) [inline src/main/java/com/scoperetail/fusion/core/application/service/command/PosterService.java:108]: Need discussion. Can we directly get the hashkey without going through the usecases?

**Review by amit-petkar — CHANGES_REQUESTED** [review]
Need discussion

File: src/main/java/com/scoperetail/fusion/core/application/service/transform/impl/DuplicateFtlTransformer.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * =====
+ */
+
+import com.scoperetail.fusion.core.application.service.transform.template.engine.FreemarkerTemplateEngine;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DuplicateFtlTransformer extends DuplicateTransformer {
```
Comment (ravi-zadpide) [inline src/main/java/com/scoperetail/fusion/core/application/service/transform/impl/DuplicateFtlTransformer.java:33]: Rename to HashKeyFtlTemplateTransformer

File: src/main/java/com/scoperetail/fusion/core/application/service/transform/impl/DuplicateTransformer.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * =====
+ */
+
+import com.scoperetail.fusion.core.application.service.transform.AbstractTransformer;
+import com.scoperetail.fusion.core.application.service.transform.template.engine.TemplateEngine;
+import com.scoperetail.fusion.core.common.HashUtil;
+
+import java.util.Map;
+
+public class DuplicateTransformer extends AbstractTransformer {
```
Comment (ravi-zadpide) [inline src/main/java/com/scoperetail/fusion/core/application/service/transform/impl/DuplicateTransformer.java:35]: make this abstract and Rename to AbstractHashKeyTemplateTransformer

File: src/main/java/com/scoperetail/fusion/core/application/service/transform/impl/DuplicateVelocityTransformer.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * =====
+ */
+
+import com.scoperetail.fusion.core.application.service.transform.template.engine.VelocityTemplateEngine;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DuplicateVelocityTransformer extends DuplicateTransformer {
```
Comment (ravi-zadpide) [inline src/main/java/com/scoperetail/fusion/core/application/service/transform/impl/DuplicateVelocityTransformer.java:33]: Rename to HashKeyVelocityTemplateTransformer

**Review by ravi-zadpide — CHANGES_REQUESTED** [review]

File: src/main/java/com/scoperetail/fusion/core/adapter/in/web/command/AbstractBaseDelegate.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+import com.scoperetail.fusion.core.adapter.in.web.BaseDelegate;
+import com.scoperetail.fusion.shared.kernel.common.annotation.UseCase;
+import lombok.AllArgsConstructor;
+
+@UseCase
+@AllArgsConstructor
+public abstract class AbstractBaseDelegate {
+  private static final String TEMPLATE_NAME = "hash_key";
```
[the remark this answers is not in the document]
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/adapter/in/web/command/AbstractBaseDelegate.java:36]: fixed

File: src/main/java/com/scoperetail/fusion/core/adapter/in/web/BaseDelegate.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ * =====
+ */
+
+public interface BaseDelegate {
```
[the remark this answers is not in the document]
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/adapter/in/web/BaseDelegate.java:29]: renamed

File: src/main/java/com/scoperetail/fusion/core/adapter/in/web/command/BaseDelegateService.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+import lombok.AllArgsConstructor;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+@UseCase
+@AllArgsConstructor
+public class BaseDelegateService implements BaseDelegate {
```
[the remark this answers is not in the document]
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/adapter/in/web/command/BaseDelegateService.java:46]: renamed

File: src/main/java/com/scoperetail/fusion/core/application/service/transform/impl/DuplicateFtlTransformer.java (hunk 2 above)
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/application/service/transform/impl/DuplicateFtlTransformer.java:33]: renamed

File: src/main/java/com/scoperetail/fusion/core/application/service/transform/impl/DuplicateVelocityTransformer.java (hunk 4 above)
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/application/service/transform/impl/DuplicateVelocityTransformer.java:33]: renamed

File: src/main/java/com/scoperetail/fusion/core/adapter/in/web/command/BaseDelegateService.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+	 @Override
+	 public boolean isNotDuplicate(final String eventName, final String templateName, final Object domainEntity) {
+		 Map<String, Object> paramsMap = new HashMap<>();
+		 paramsMap.put(Transformer.DOMAIN_ENTITY, domainEntity);
+		 String hashKey = null;
+		 Optional<UseCaseConfig> useCase = fusionConfig.getUsecases().stream().filter(u -> u.getName().equals(eventName))
+				 .findFirst();
+
+		 if (useCase.isPresent() && useCase.get().getDedupeCheck()) {
```
[the remark this answers is not in the document]
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/adapter/in/web/command/BaseDelegateService.java:60]: done

**Review by amit-petkar — APPROVED** [review]
Approved.

**Pull request merged**